tstate/tstate_view.go

```diff
@@ -39,8 +40,10 @@ type TStateView struct {
 	// operations allows for reverting state to a certain point-in-time.
 	ops []*op
 
-	// We don't differentiate between read and write scope.
-	scope        set.Set[string] // stores a list of managed keys in the TState struct
+	// stores a list of managed keys in the TState struct
+	readScope    set.Set[string]
+	writeScope   set.Set[string]
+	rwScope      set.Set[string]
```

Comment:
I would just have a single `map[string]Mode` of our state keys, and just look up the key's bit value in its `Mode` and get modify `checkScope` to take in a parameter of a key + the perm to look up instead of maintaining separate maps for every permutation of scopes.